apport/crashdb_impl/github.py

```diff
@@ -0,0 +1,316 @@
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 2 of the License, or (at your
+# option) any later version.  See http://www.gnu.org/copyleft/gpl.html for
+# the full text of the license.
+
+import json
+import time
+from dataclasses import dataclass
+
+import requests
+
+import apport
+import apport.crashdb
+
+
+class Github:
+    """Wrapper around Github API, used to log in and post issues."""
+
+    __last_request: float = time.time()
+
+    def __init__(self, client_id, message_callback):
+        self.__client_id = client_id
+        self.__authentication_data = None
+        self.__access_token = None
+        self.__cooldown = None
+        self.__expiry = None
+        self.message_callback = message_callback
+
+    @staticmethod
+    def _stringify(data: dict) -> str:
+        """Takes a dict and returns it as a string for POSTing."""
+        string = ""
+        for key, value in data.items():
+            string = f"{string}&{key}={value}"
+        return string
+
+    def _post(self, url: str, data: str):
+        """Posts the given data to the given URL.
+        Uses auth token if available"""
+        headers = {"Accept": "application/vnd.github.v3+json"}
+        if self.__access_token:
+            headers["Authorization"] = f"token {self.__access_token}"
+        try:
+            result = requests.post(
+                url, headers=headers, data=data, timeout=5.0
+            )
+        except requests.RequestException as err:
+            self.message_callback(
+                "Failed connection",
+                f"Failed connection to {url}.\n"
+                + "Please check your internet connection and try again.",
+            )
+            raise err
+        finally:
+            self.__last_request = time.time()
+
+        result.raise_for_status()  # Not using UI: the user can't do much here
+        return json.loads(result.text)
+
+    def api_authentication(self, url: str, data: dict):
+        return self._post(url, self._stringify(data))
+
+    def api_open_issue(self, owner: str, repo: str, data: dict):
+        url = f"https://api.github.com/repos/{owner}/{repo}/issues"
+        return self._post(url, json.dumps(data))
+
+    def __enter__(self):
+        """Enters login process. At exit, login process ends."""
+        data = {"client_id": self.__client_id, "scope": "public_repo"}
+        url = "https://github.com/login/device/code"
+        response = self.api_authentication(url, data)
+
+        prompt = "Posting an issue requires a Github account. If you have "
+        prompt += "one, please follow these steps to log in.\n"
+        prompt += "\n"
+        prompt += "Open the following URL. When requested, write this code "
+        prompt += "to enable apport to open an issue.\n"
+        prompt += "URL:  {url}\n"
+        prompt += "Code: {code}"
```

Comment:
```suggestion
        prompt = (
            "Posting an issue requires a Github account. If you have "
            "one, please follow these steps to log in.\n"
            "\n"
            "Open the following URL. When requested, write this code "
            "to enable apport to open an issue.\n"
            "URL:  {url}\n"
            "Code: {code}"
        )
```

Reply:
Or even use f-strings:
```suggestion
        prompt = (
            f"Posting an issue requires a Github account. If you have "
            f"one, please follow these steps to log in.\n"
            f"\n"
            f"Open the following URL. When requested, write this code "
            f"to enable apport to open an issue.\n"
            f"URL:  {response['verification_uri']}\n"
            f"Code: {response['user_code']}"
        )
```